core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -95,30 +107,84 @@ public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand
+               continue;
+            }
+            Set<Integer> segments = ch.getPrimarySegmentsForOwner(member);
+            if (!segments.isEmpty()) {
+               Map<Object, Object> segmentEntriesMap = 
+                     new ReadOnlySegmentAwareMap<Object, Object>(originalMap, ch, 
+                           segments);
+               if (!segmentEntriesMap.isEmpty()) {
+                  PutMapCommand copy = new PutMapCommand(command);
+                  copy.setMap(segmentEntriesMap);
+                  NotifyingNotifiableFuture<Object> future = new NotifyingFutureImpl<Object>();
+                  rpcManager.invokeRemotelyInFuture(Collections.singletonList(member), 
+                        copy, options, future);
+                  futures.add(future);
+               }
+            }
          }
-         primaryOwners.remove(rpcManager.getAddress());
-         if (!primaryOwners.isEmpty()) {
-            rpcManager.invokeRemotely(primaryOwners, command, rpcManager.getDefaultRpcOptions(isSynchronous(command)));
+         if (futures.size() > 0) {
+            CompositeNotifyingFuture<Object> compFuture = 
+                  new CompositeNotifyingFuture<Object>(futures);
+            try {
+               compFuture.get(options.timeout(), TimeUnit.MILLISECONDS);
+            } catch (TimeoutException e) {
+               throw new CacheException(e);
+            }
          }
       }
 
-      if (!command.isForwarded()) {
-         //I need to forward this to all the nodes that are secondary owners
-         Set<Object> keysIOwn = new HashSet<Object>(command.getAffectedKeys().size());
-         for (Object k : command.getAffectedKeys()) {
-            if (cdl.localNodeIsPrimaryOwner(k)) {
-               keysIOwn.add(k);
+      if (!command.isForwarded() && ch.getNumOwners() > 1) {
+         // Now we find all the segments that we own and map our backups to those
+         Map<Address, Set<Integer>> backupOwnerSegments = new HashMap<>();
+         int segmentCount = ch.getNumSegments();
+         for (int i = 0; i < segmentCount; ++i) {
+            Iterator<Address> iter = ch.locateOwnersForSegment(i).iterator();
+
+            if (iter.next().equals(localAddress)) {
+               while (iter.hasNext()) {
+                  Address backupOwner = iter.next();
+                  Set<Integer> segments = backupOwnerSegments.get(backupOwner);
+                  if (segments == null) {
+                     backupOwnerSegments.put(backupOwner, (segments = new HashSet<>()));
+                  }
+                  segments.add(i);
+               }
             }
          }
-         Collection<Address> backupOwners = cdl.getOwners(keysIOwn);
-         if (backupOwners == null || !backupOwners.isEmpty()) {
+
+         int backupOwnerSize = backupOwnerSegments.size();
+         if (backupOwnerSize > 0) {
+            List<NotifyingFuture<Object>> futures = new ArrayList<NotifyingFuture<Object>>(
```

Comment:
Ah yes this is for backups... somehow the code must have gotten removed.  I am guessing this is why the tests are failing.  I hope to make a PR soon for it.